File: app/src/main/java/knf/kuma/videoservers/Headers.java
package knf.kuma.videoservers;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Pair;

import java.util.ArrayList;
import java.util.List;

public class Headers implements Parcelable {
    public static final Creator<Headers> CREATOR = new Creator<Headers>() {
        @Override
        public Headers createFromParcel(Parcel in) {
            return new Headers(in);
        }

        @Override
        public Headers[] newArray(int size) {
            return new Headers[size];
        }
    };
    private List<Pair<String, String>> headers = new ArrayList<>();
    private List<Pair<String, String>> cookies = new ArrayList<>();

    public Headers() {
    }

    public Headers(Parcel parcel) {
        headers = parcel.readArrayList(null);
        cookies = parcel.readArrayList(null);
    }

    public void addHeader(String key, String value) {
        headers.add(new Pair<>(key, value));
    }

    public void addCookie(String key, String value) {
        cookies.add(new Pair<>(key, value));
    }

    public List<Pair<String, String>> getHeaders() {
        if (cookies.size() > 0)
            headers.add(new Pair<>("Cookie", getCookies()));
        return headers;
    }

    private String getCookies() {
        StringBuilder builder = new StringBuilder();
        for (Pair<String, String> pair : cookies)
            builder.append(pair.first)
                    .append('=')
                    .append(pair.second)
                    .append("; ");
        return builder.toString().trim();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeList(headers);
        parcel.writeList(cookies);
    }
}

File: app/src/main/java/knf/kuma/videoservers/Headers.kt
package knf.kuma.videoservers

import android.os.Parcel
import android.os.Parcelable
import knf.kuma.commons.noCrash

class Headers : Parcelable {
    var headers: MutableList<Pair<String, String>> = mutableListOf()
    private var cookies: MutableList<Pair<String, String>> = mutableListOf()

    constructor()

    constructor(block: MutableList<Pair<String, String>>.() -> Unit) {
        block(headers)
    }

    constructor(parcel: Parcel) {
        headers = parcel.readArrayList(null)?.filterIsInstance<Pair<String, String>>() as MutableList<Pair<String, String>>
        cookies = parcel.readArrayList(null)?.filterIsInstance<Pair<String, String>>() as MutableList<Pair<String, String>>
    }

    fun createHeadersList(): List<String> = mutableListOf<String>().apply {
        headers.forEach {
            add(it.first)
            add(it.second)
        }
        if (cookies.size > 0) {
            add("Cookie")
            add(getCookies())
        }
    }

    fun createHeaders(): List<Pair<String, String>> =
            mutableListOf<Pair<String, String>>().apply {
                addAll(headers)
                if (cookies.size > 0)
                    add("Cookie" to getCookies())
            }

    fun createHeadersMap(): HashMap<String, String> {
        val map = HashMap<String, String>()
        createHeaders().forEach {
            map[it.first] = it.second
        }
        return map
    }

    fun addHeader(key: String, value: String) {
        headers.add(Pair(key, value))
    }

    fun addCookie(key: String, value: String) {
        cookies.add(Pair(key, value))
    }

    fun getCookies(): String {
        val builder = StringBuilder()
        for (pair in cookies)
            builder.append(pair.first)
                    .append('=')
                    .append(pair.second)
                    .append("; ")
        return builder.toString().trim { it <= ' ' }
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        noCrash {
            parcel.writeList(headers as List<*>?)
            parcel.writeList(cookies as List<*>?)
        }
    }

    override fun toString(): String {
        return createHeadersList().toString()
    }

    companion object CREATOR : Parcelable.Creator<Headers> {
        override fun createFromParcel(parcel: Parcel): Headers {
            return Headers(parcel)
        }

        override fun newArray(size: Int): Array<Headers?> {
            return arrayOfNulls(size)
        }
    }
}
